programs/stake/src/stake_state.rs

```diff
@@ -1158,6 +1144,135 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+        // succeeds. Considering it here would needlessly prevent merging stake
+        // accounts with differing data lengths, which already exist in the wild
+        // due to an SDK bug
+        if stake.authorized == source.authorized && stake.lockup == source.lockup {
+            Ok(())
+        } else {
+            Err(StakeError::MergeMismatch.into())
+        }
+    }
+
+    fn active_delegations_can_merge(
+        stake: &Delegation,
+        source: &Delegation,
+    ) -> Result<(), InstructionError> {
+        if stake.voter_pubkey == source.voter_pubkey
+            && (stake.warmup_cooldown_rate - source.warmup_cooldown_rate).abs() < f64::EPSILON
+            && stake.deactivation_epoch == Epoch::MAX
+            && source.deactivation_epoch == Epoch::MAX
+        {
+            Ok(())
+        } else {
+            Err(StakeError::MergeMismatch.into())
+        }
+    }
+
+    fn active_stakes_can_merge(stake: &Stake, source: &Stake) -> Result<(), InstructionError> {
+        Self::active_delegations_can_merge(&stake.delegation, &source.delegation)?;
+        // `credits_observed` MUST match to prevent earning multiple rewards
+        // from a stake account by merging it into another stake account that
+        // is small enough to not be paid out every epoch. This would effectively
+        // reset the larger stake accounts `credits_observed` to that of the
+        // smaller account.
+        if stake.credits_observed == source.credits_observed {
+            Ok(())
+        } else {
+            Err(StakeError::MergeMismatch.into())
+        }
+    }
+
+    fn merge(self, source: Self) -> Result<Option<StakeState>, InstructionError> {
+        Self::metas_can_merge(self.meta(), source.meta())?;
+        self.stake()
+            .zip(source.stake())
+            .map(|(stake, source)| Self::active_stakes_can_merge(stake, source))
+            .unwrap_or(Ok(()))?;
+        let merged_state = match (self, source) {
+            (Self::Inactive(_, _), Self::Inactive(_, _)) => None,
+            (Self::Inactive(_, _), Self::ActivationEpoch(_, _)) => None,
+            (Self::ActivationEpoch(meta, mut stake), Self::Inactive(_, source_lamports)) => {
+                stake.delegation.stake += source_lamports;
+                Some(StakeState::Stake(meta, stake))
+            }
+            (
+                Self::ActivationEpoch(meta, mut stake),
+                Self::ActivationEpoch(source_meta, source_stake),
+            ) => {
+                let source_lamports =
+                    source_meta.rent_exempt_reserve + source_stake.delegation.stake;
+                stake.delegation.stake += source_lamports;
+                Some(StakeState::Stake(meta, stake))
+            }
+            (Self::FullyActive(meta, mut stake), Self::FullyActive(_, source_stake)) => {
+                // Don't stake the source account's `rent_exempt_reserve` to
+                // protect against the magic activation loop hole. It will
```

Comment:
Super nit-picky: loophole is one word (cc: hyphenated use on L1048)

Reply:
TFW you're unsure and too lazy to look it up... Use them all! :upside_down_face: